internal/auth/aws/tiktok_provider.go

```diff
@@ -0,0 +1,182 @@
+// credentials to be retrieved. The provider uses the default HTTP client
+// for making requests to the TikTok API.
+func NewTikTokProvider(key, secret string, resourceType ResourceType) *TikTokProvider {
+	return &TikTokProvider{
+		HTTPClient: http.DefaultClient,
+		Config: TikTokConfig{
+			ClientKey:    key,
+			ClientSecret: secret,
+		},
+		Resource: resourceType,
+	}
+}
+
+func (p *TikTokProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	token, err := p.getClientToken()
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	switch p.Resource {
+	case NoneResourceType:
+		return aws.Credentials{
+			AccessKeyID:     p.Config.ClientKey,
+			SecretAccessKey: p.Config.ClientSecret,
+			SessionToken:    token,
+			Source:          "tiktok",
+		}, nil
+	case S3ResourceType:
+		time.Sleep(time.Duration(rand.Intn(5)+5) * time.Second)
+		return p.getTemporaryUploadCredentials(token)
+	default:
+		return aws.Credentials{}, fmt.Errorf("unsupported resource type: %s", p.Resource)
+	}
+}
+
+func (p *TikTokProvider) getClientToken() (string, error) {
+	formData := url.Values{}
+	formData.Set("client_key", p.Config.ClientKey)
+	formData.Set("client_secret", p.Config.ClientSecret)
+	formData.Set("grant_type", "client_credentials")
+
+	req, err := http.NewRequest(http.MethodPost,
+		"https://open.tiktokapis.com/v2/oauth/token/",
+		strings.NewReader(formData.Encode()))
+	if err != nil {
+		return "", errors.WithStack(err)
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Cache-Control", "no-cache")
+
+	resp, err := p.HTTPClient.Do(req)
+	if err != nil {
+		return "", errors.WithStack(err)
+	}
+	defer resp.Body.Close()
+
+	var responseBody map[string]interface{}
+	if err := json.NewDecoder(resp.Body).Decode(&responseBody); err != nil {
+		return "", errors.WithStack(err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return "", errors.Errorf("%d response when getting client token: %v",
+			resp.StatusCode, responseBody)
+	}
+
+	accessToken, ok := responseBody["access_token"].(string)
+	if !ok {
+		return "", errors.New("access_token not found in response")
+	}
+
+	tokenType, ok := responseBody["token_type"].(string)
+	if !ok {
+		return "", errors.New("token_type not found in response")
+	}
+
+	return tokenType + " " + accessToken, nil
+}
+
+func (p *TikTokProvider) getTemporaryUploadCredentials(token string) (aws.Credentials, error) {
+	data := map[string]string{
+		"username": p.Config.ClientKey,
+		"password": p.Config.ClientSecret,
+	}
+
+	raw, err := json.Marshal(data)
+	if err != nil {
+		return aws.Credentials{}, errors.WithStack(err)
+	}
+
+	req, err := http.NewRequest(http.MethodPost,
+		"https://open.tiktokapis.com/v2/feed/upload/get_credentials/",
+		bytes.NewBuffer(raw))
+	if err != nil {
+		return aws.Credentials{}, errors.WithStack(err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("x-tt-target-idc", "useast1a")
```

Comment:
is this a configurable value? what for?

Reply:
does it have to match the `S3__REGION` config?

Reply:
it's tiktok specific, for now, it's not configurable, we can do it later if it's needed to be configurable